C/Cpp_include/c4Index.hh

```diff
@@ -0,0 +1,92 @@
+//
+// c4Index.hh
+//
+// Copyright 2024-Present Couchbase, Inc.
+//
+// Use of this software is governed by the Business Source License included
+// in the file licenses/BSL-Couchbase.txt.  As of the Change Date specified
+// in that file, in accordance with the Business Source License, use of this
+// software will be governed by the Apache License, Version 2.0, included in
+// the file licenses/APL2.txt.
+//
+
+#pragma once
+#include "c4Base.hh"
+#include "c4IndexTypes.h"
+#include "fleece/InstanceCounted.hh"
+
+C4_ASSUME_NONNULL_BEGIN
+
+// ************************************************************************
+// This header is part of the LiteCore C++ API.
+// If you use this API, you must _statically_ link LiteCore;
+// the dynamic library only exports the C API.
+// ************************************************************************
+
+
+/** Represents an index. Acts as a factory for C4IndexUpdater objects. */
+struct C4Index
+    : public fleece::RefCounted
+    , public fleece::InstanceCountedIn<C4Index>
+    , C4Base {
+    C4Collection* getCollection() const { return _collection; }
+
+    slice getName() const { return _name; }
+
+#ifdef COUCHBASE_ENTERPRISE
+    /// Finds new or updated documents for which vectors need to be recomputed by the application.
+    /// If there are none, returns NULL. */
+    Retained<struct C4IndexUpdater> beginUpdate(size_t limit);
+#endif
+
+  protected:
+    friend class litecore::CollectionImpl;
+    static Retained<C4Index> getIndex(C4Collection*, slice name);
+
+    Retained<C4Collection> _collection;
+    std::string            _name;
+};
+
+#ifdef COUCHBASE_ENTERPRISE
+
+/** Describes a set of index values that need to be computed by the application,
+    to update a lazy index after its Collection has changed.
+    You should:
+    1. Call `valueAt` for each of the `count` items to get the Fleece value, and:
+      1.1. Compute a vector from this value
+      1.2. Call `setVectorAt` with the resulting vector, or with nullptr if none.
+    2. Finally, open a transaction and call `finish` to apply the updates to the index. */
+struct C4IndexUpdater final
+    : public fleece::RefCounted
+    , public fleece::InstanceCountedIn<C4IndexUpdater>
+    , C4Base {
+    /// The number of vectors to compute.
+    size_t count() const;
+
+    /// Returns the i'th value to compute a vector from.
+    /// This is the value of the expression in the index spec.
+    FLValue valueAt(size_t i) const;
+
+    /// Sets the vector for the i'th value. If you don't call this, it's assumed there is no
+    /// vector, and any existing vector will be removed upon `finish`.
+    void setVectorAt(size_t i, const float* vector, size_t dimension);
```

Comment:
So for the case that the remote ML model is offline (e.g. no internet connection) during the updating process,  do I understand correctly that users will not call this function so the docs will be included again next time?

Reply:
This iteration of the implementation doesn't support temporarily skipping a document; it just uses a `lastSequence` number to remember which documents are up to date. I'll create another PR after this one to support it.

In this PR I'll add a `skipVectorAt()` method to make it explicit that a vector is being skipped.